pkg/collector/sharedlibrary/ffi.c

```diff
@@ -0,0 +1,137 @@
+		*error = strdup(error_msg);
+		return lib_handles;
+    }
+
+    // get pointer of 'Run' symbol
+    lib_handles.run = (run_function_t *)GetProcAddress(lib_handles.lib, "Run");
+    if (!lib_handles.run) {
+        char error_msg[256];
+        int error_code = GetLastError();
+        snprintf(error_msg, sizeof(error_msg), "unable to get shared library 'Run' symbol, error code: %d", error_code);
+		*error = strdup(error_msg);
+		return lib_handles;
+    }
+
+    // try to get pointer of 'Version' symbol (not required)
+    lib_handles.version = (version_function_t *)GetProcAddress(lib_handles.lib, "Version");
+
+    return lib_handles;
+}
+
+void close_shared_library(void *lib_handle, const char **error) {
+	// verify pointer
+	if (!lib_handle) {
+		*error = strdup("pointer to shared library is NULL");
+        return;
+	}
+
+    // close the library and check for errors (error_code == 0)
+    int error_code = FreeLibrary(lib_handle);
+    if (!error_code) {
+        char error_msg[256];
+        int error_code = GetLastError();
+        snprintf(error_msg, sizeof(error_msg), "unable to close shared library, error code: %d", error_code);
+		*error = strdup(error_msg);
+    }
+}
+
+#else
+
+handles_t load_shared_library(const char *lib_path, const char **error) {
+    handles_t lib_handles;
+    char *dlsym_error = NULL;
+
+    // load the library
+    lib_handles.lib = dlopen(lib_path, RTLD_LAZY | RTLD_GLOBAL);
+    
+    // catch library opening error
+    dlsym_error = dlerror();
+    if (dlsym_error) {
+		*error = strdup(dlsym_error);
+		return lib_handles;
+    }
+
+    // get pointer of 'Run' symbol
+    lib_handles.run = (run_function_t *)dlsym(lib_handles.lib, "Run");
+    
+    // catch symbol errors and close the library if there are any
+    dlsym_error = dlerror();
+    if (dlsym_error) {
+		dlclose(lib_handles.lib);
+		*error = strdup(dlsym_error);
+		return lib_handles;
+    }
+
+    // try to get pointer of 'Version' symbol (not required)
+    lib_handles.version = (version_function_t *)dlsym(lib_handles.lib, "Version");
+    
+    // set the pointer to NULL is there's an error, the 'Version' symbol is not required
+    dlsym_error = dlerror();
+    if (dlsym_error) {
+		lib_handles.version = NULL;
+    }
+
+    return lib_handles;
+}
+
+void close_shared_library(void *lib_handle, const char **error) {
+    // verify library handle
+	if (!lib_handle) {
+        *error = strdup("pointer to shared library is NULL");
+        return;
+	}
+
+    // close the library
+    dlclose(lib_handle);
+
+    // check for closing errors
+    char *dlsym_error = dlerror();
+    if (dlsym_error) {
+        *error = strdup(dlsym_error);
+    }
+}
+#endif
+
+void run_shared_library(run_function_t *run_ptr, char *check_id, char *init_config, char *instance_config, aggregator_t *aggregator, const char **error) {
+    // verify `Run` pointer
+    if (!run_ptr) {
+        *error = strdup("pointer to 'Run' symbol of the shared library is NULL");
```

Comment:
It's a shame to `strdup` constant strings just so that we can free them later, if we really wanted to avoid allocations we could do something `errno` -like with a static list of known errors
I don't think it's necessary for now, the code is simpler as is and it's not the happy path